pyodide-build/pyodide_build/buildpkg.py

```diff
@@ -106,129 +134,240 @@ def check_checksum(path: Path, pkg: Dict[str, Any]):
         fd.write(b"\n")
 
 
-def compile(path: Path, srcpath: Path, pkg: Dict[str, Any], args, bash_runner):
+def compile(
+    pkg_root: Path,
+    srcpath: Path,
+    build_metadata: Dict[str, Any],
+    bash_runner: BashRunnerWithSharedEnvironment,
+    *,
+    target: str,
+    install_dir: str,
+):
+    """
+    Runs pywasmcross for the package. The effect of this is to first run setup.py
+    with compiler wrappers subbed in, which don't actually build the package but
+    write the compile commands to build.log. Then we walk over build log and invoke
+    the same set of commands but with some flags munged around or removed to make it
+    work with emcc.
+
+    In any case, only works for Python packages, not libraries or shared libraries
+    which don't have a setup.py.
+
+    Parameters
+    ----------
+    pkg_root
+        The path to the root directory for the package. Generally
+        $PYODIDE_ROOT/packages/<PACKAGES>
+
+    srcpath
+        The path to the source. We extract the source into the build directory, so it
+        will be something like
+        $(PYOIDE_ROOT)/packages/<PACKAGE>/build/<PACKAGE>-<VERSION>.
+
+    build_metadata
+        The build section from meta.yaml.
+
+    bash_runner
+        The runner we will use to execute our bash commands. Preserves environment
+        variables from one invocation to the next.
+
+    target
+        The path to the target Python installation
+
+    install_dir
+        Directory for installing built host packages. Defaults to setup.py
+        default. Set to 'skip' to skip installation. Installation is
+        needed if you want to build other packages that depend on this one.
+    """
     if (srcpath / ".built").is_file():
         return
 
-    orig_dir = Path.cwd()
-    os.chdir(srcpath)
-    if pkg.get("build", {}).get("skip_host", True):
+    if build_metadata.get("skip_host", True):
         bash_runner.env["SKIP_HOST"] = ""
 
-    try:
+    replace_libs = ";".join(build_metadata.get("replace-libs", []))
+
+    with chdir(srcpath):
         subprocess.run(
             [
                 sys.executable,
                 "-m",
                 "pyodide_build",
                 "pywasmcross",
                 "--cflags",
-                args.cflags + " " + pkg.get("build", {}).get("cflags", ""),
+                build_metadata["cflags"],
                 "--cxxflags",
-                args.cxxflags + " " + pkg.get("build", {}).get("cxxflags", ""),
+                build_metadata["cxxflags"],
                 "--ldflags",
-                args.ldflags + " " + pkg.get("build", {}).get("ldflags", ""),
+                build_metadata["ldflags"],
                 "--target",
-                args.target,
+                target,
                 "--install-dir",
-                args.install_dir,
+                install_dir,
                 "--replace-libs",
-                ";".join(pkg.get("build", {}).get("replace-libs", [])),
+                replace_libs,
```

Comment:
Minor point, but there is a `cwd` option to subprocess.run without would avoid using our own wrapper for this.